src/nvim/api/ui.c

```diff
@@ -340,7 +345,62 @@ void nvim_ui_pum_set_height(uint64_t channel_id, Integer height, Error *err)
                   "It must support the ext_popupmenu option");
     return;
   }
-  ui->pum_height = (int)height;
+
+  ui->pum_nlines = (int)height;
+}
+
+/// Tells Nvim the geometry of the popumenu, to align floating
+/// windows with an external popup menu. Note that this method
+/// is not to be confused with |nvim_ui_pum_set_height()|, which
+/// sets the number of visible items in the popup menu, while
+/// this function sets the bounding box of the popup menu,
+/// including visual decorations such as boarders and sliders.
+/// Floats need not use the same font size, nor be anchored to
+/// exact grid corners, so one can set floating-point numbers
+/// to the popup menu geometry.
+///
+/// @param channel_id
+/// @param width   Popupmenu width.
+/// @param height  Popupmenu height.
+/// @param row     Popupmenu row.
+/// @param col     Popupmenu height.
+/// @param[out] err Error details, if any.
+void nvim_ui_pum_set_bounds(uint64_t channel_id, Float width, Float height,
+                            Float row, Float col, Error *err)
+  FUNC_API_SINCE(7) FUNC_API_REMOTE_ONLY
+{
+  if (!pmap_has(uint64_t)(connected_uis, channel_id)) {
+    api_set_error(err, kErrorTypeException,
+                  "UI not attached to channel: %" PRId64, channel_id);
+    return;
+  }
+
+  UI *ui = pmap_get(uint64_t)(connected_uis, channel_id);
+  if (!ui->ui_ext[kUIPopupmenu]) {
+    api_set_error(err, kErrorTypeValidation,
+                  "UI must support the ext_popupmenu option");
+    return;
+  }
+
+  if (row < 0) {
+    api_set_error(err, kErrorTypeValidation, "Expected pumpos row >= 0");
```

Comment:
can remove "pumpos" here and in the other cases.

Reply:
Though even the `row < 0` and `col < 0` checks sholud be removed. In a multigrid UI floats and the pum can float outside the border of the main window, including above/left of it.

Reply:
updated.